components/omega/doc/design/Halo.md

```diff
@@ -0,0 +1,260 @@
+
+### 2.1 Requirement: Compatible with all types of arrays
+
+Handle halo exchanges for arrays of each supported data type and dimensionality.
+
+### 2.2 Requirement: Exchange at all types of mesh elements
+
+Meshes are composed of three types of elements (cells, edges, and vertices) 
+at which values can be defined, we must be able to handle exchanges at each 
+type of element.
+
+### 2.3 Requirement: Variable halo depth
+
+The default minimum required halo depth for most of the algorithms we will 
+use is three cells. We should also be able to handle different halo depths.
+
+### 2.4 Requirement: Exchange lists
+
+Each MPI rank needs a send list and receive list for every neighboring rank and 
+each type of mesh element (cells, edges, and vertices) containing the local
+indices for each owned element to be packed and sent to neighboring ranks and 
+each element owned by neighboring ranks to be received and unpacked into local 
+field arrays. Exchange lists should be organized by halo layer.
+
+### 2.5 Requirement: Send and receive buffers
+
+Each MPI rank needs two arrays for every neighboring rank to serve as a send 
+buffer and a receive buffer to be passed to the MPI send and receive routines.
+Only one buffer needs to be communicated to each neighbor at a time, so the 
+buffers can be persistent and reused for each exchange to avoid allocating and 
+deallocating new buffer arrays for each exchange. The buffer size would 
+therefore need to be set to the largest necessary buffer length and only a 
+subset of the buffer array needs to be communicated for smaller exchanges. 
+
+### 2.6 Requirement: Non-blocking communication
+
+To efficiently handle exchanges between each rank and all of its neighbors, we
+will use the non-blocking MPI routines ```MPI_Isend``` and ```MPI_Irecv```. The 
+```MPI_Test``` routine will be used to determine when each communication is 
+completed. Boolean variables will be needed to track when messages have been 
+received and when buffers have been unpacked.
+
+### 2.7 Desired: Communicate subset of halo layers
+
+We may want the ability to communicate individual halo layers as opposed to
+always exchanging the entire halo for a given field. Exchange lists should be 
+organized by halo layer to facilitate this.
+
+### 2.8 Desired: Exchange multiple fields simultaneously
+
+Communications in MPAS are generally latency limited, so it may be advantageous 
+to exchange multiple fields at the same time, packing the halo elements of 
+multiple fields into the same buffer to reduce the number of messages.
+
+### 2.9 Desired: Multiple environments/decompositions
+ 
+OMEGA runs may contain multiple communication environments or mesh 
+decompositions, would need to be able to perform halo exchanges in these 
+circumstances.
+
+### 2.10 Desired: OpenMP threading
+
+If we allow OpenMP threading, halo exchanges would be carried out by the master 
+thread, and a process to carry out local exchanges would be needed.
+
```

Comment:
I suspect we will indeed stick with thread master doing this but there are other threading approaches that could work. So maybe just change this to say that the implementation is thread-safe.